### PR TITLE
test: add layerwise remote backend tests and fix bytes_read guard

### LINKED ISSUE #2752: layerwise is not working on remote backend
layerwise=true is not working together with remote backend at all.
not sure if this is on purpose or not , but below is the error log on lm & redis backends:

lm remote backend error log: 

LMCache ERROR: Put task failed for key LayerCacheEngineKey(model_name='/workspace/Meta-Llama-3-8B-Instruct/', world_size=1, worker_id=0, chunk_hash=4132912831621080023, dtype=torch.bfloat16, request_configs=None, tags=None, _dtype_str='bfloat16', layer_id=11): Shape dimension should be 4 (remote_backend.py:196:lmcache.v1.storage_backend.remote_backend)

redis remote backend error log:

LMCache WARNING: batched put error: Shape dimension should be 4 (instrumented_connector.py:186:lmcache.v1.storage_backend.connector.instrumented_connector)

### PR DESCRIPTION
**Pull request by ianliuy**
## Summary

Follow-up to #2751 which fixes the core `assert len(shape) == 4` crashes for layerwise remote backend support.

This PR adds missing test coverage and fixes subtle bugs found during review.

### Fix: `bytes_read <= 0` guard in `reshape_partial_chunk`

The original guard used `bytes_read == 0`, but Python's modulo operator returns non-negative results for positive divisors:

```python
-256 % 256 == 0   # True -- bypasses the guard!
-256 > 81920      # False -- bypasses the check!
```

A negative `bytes_read` would silently produce `shape_list[token_dim] = -1` and truncate data from the wrong end. Changed to `bytes_read <= 0`.

### Fix: Multi-group `reshape_partial_chunk` support

`reshape_partial_chunk` previously only updated `meta.shape` (singular) and forced `group_prefix_sum = [0, bytes_read]`. For multi-group MemoryObj (e.g., separate K and V tensors), this left `meta.shapes` stale and broke `get_tensor(1)`. Now:
- All shapes in `meta.shapes` are updated (token dim adjusted per shape)
- `group_prefix_sum` is recalculated from actual `shapes x dtypes`
- Falls back to `[0, bytes_read]` when `shapes`/`dtypes` are None

### Fix: `assert`  `ValueError` in `pad_shape_to_4d`

Replaced `assert len(shape) <= 4` with `raise ValueError(...)` so the check is never disabled by `python -O`.

### Tests: `reshape_partial_chunk` with 2D/3D/4D shapes

`reshape_partial_chunk` had zero test coverage for sub-4D shapes (the exact shapes introduced by layerwise mode). Added parametrized tests covering:
- 4D: `[kv, layers, tokens, hidden]` (non-layerwise)
- 3D: `[tokens, 2, hidden]` (layerwise MHA, `token_dim=0`)
- 2D: `[tokens, hidden]` (layerwise MLA, `token_dim=0`)
- Invalid/edge cases: zero bytes, misaligned, over-size, negative

### Tests: `ClientMetaMessage` / `ServerMetaMessage` sub-4D round-trips

Added serialization round-trip tests for 3D and 2D shapes through the wire protocol, complementing the tests added in #2751 for `RemoteMetadata`.

## Relationship to #2751

This PR is designed to be merged **after** #2751. It merges #2751's branch as a base and adds improvements on top.

Closes #2752

---

> [!NOTE]
> **Medium Risk**
> Touches remote-backend partial-read reshaping logic and multi-group size accounting; mistakes here could corrupt tensor views or sizes when reading from remote storage. Changes are well-scoped and backed by new parametrized tests, reducing regression risk.
> 
> **Overview**
> Improves remote-backend handling of *partial chunk reads* by tightening validation (`bytes_read <= 0`) and correctly recomputing token counts for layerwise 2D/3D shapes (token dimension selection no longer relies on the stored full-chunk `fmt`).
> 
> Extends `reshape_partial_chunk` to update multi-group metadata (`meta.shapes`) and to recompute `group_prefix_sum` from `shapes`×`dtypes` so `get_size()`/`get_tensor()` reflect the truncated buffer.
> 
> Hardens protocol shape serialization by replacing an `assert` in `pad_shape_to_4d` with a `ValueError`, and adds tests for sub-4D shape round-trips through `RemoteMetadata`, `ClientMetaMessage`, and `ServerMetaMessage`, plus new connector tests covering 2D/3D/4D layouts and invalid `bytes_read` cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 4dfb87c8571b238b6038f20b234267f53a36aa16. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: lmcache/v1/protocol.py [hunk 1]
```diff
@@ -96,6 +96,50 @@ def get_remote_metadata_bytes():
...
+
+    Returns:
+        A list of exactly 4 integers representing the padded shape.
+
+    Raises:
+        AssertionError: If the shape has more than 4 dimensions.
+    """
+    assert len(shape) <= 4, (
+        f"Shape dimension must be <= 4 for serialization, got {len(shape)}"
+    )
```
[the remark this answers is not in the document]
Comment (ianliuy) [reply lmcache/v1/protocol.py]: Agreed  replaced `assert len(shape) <= 4` with `raise ValueError(...)` in the latest commit. Production code shouldn't rely on assertions that can be disabled with `-O`. Thanks!

File: lmcache/v1/storage_backend/connector/base_connector.py [hunk 2]
```diff
@@ -107,10 +107,29 @@ def reshape_partial_chunk(
...
+            # Layerwise MLA 2D: [num_tokens, hidden_dim]
+            token_dim = 0
+        shape_list[token_dim] = bytes_read // self.single_token_size
         actual_shape = torch.Size(shape_list)
         memory_obj.raw_data = memory_obj.raw_data[:bytes_read]
         memory_obj.meta.shape = actual_shape
+        # Sync group_prefix_sum so that get_size() / byte_array reflect the
+        # truncated size rather than the original full-chunk size.
+        if hasattr(memory_obj, "group_prefix_sum"):
+            memory_obj.group_prefix_sum = [0, bytes_read]  # type: ignore[attr-defined]
```
[the remark this answers is not in the document]
Comment (ianliuy) [reply lmcache/v1/storage_backend/connector/base_connector.py]: Good catch  addressed in the latest commit (4cdf8ed).

Changes:
- `meta.shapes` is now updated per-group (each shape's token dimension adjusted independently)
- `group_prefix_sum` is recalculated from actual `shapes  dtypes` rather than forced to `[0, bytes_read]`
- Uses `zip(..., strict=True)` to guard against shapes/dtypes length mismatch
- Falls back to `[0, bytes_read]` when shapes/dtypes are None (single-group path)

Thanks for the thorough review!

File: lmcache/v1/gpu_connector/gpu_connectors.py [hunk 3]
```diff
@@ -1198,12 +1199,14 @@ def batched_to_gpu(self, starts: List[int], ends: List[int], **kwargs):
 
         num_tokens = len(slot_mapping_full)
 
+        mem_fmt = MemoryFormat.KV_MLA_FMT if self.use_mla else MemoryFormat.KV_T2D
```
[the remark this answers is not in the document]
Comment (ianliuy) [reply lmcache/v1/gpu_connector/gpu_connectors.py:1202]: This is a valid observation about the broader MemoryFormat design. However, the `KV_T2D` / `KV_MLA_FMT` format assignment here comes from PR #2751's original fix, and addressing the MemoryFormat enum's dimensionality semantics for layerwise layouts would be a larger refactor beyond this PR's scope.

We've documented the mismatch in `base_connector.py` (the NOTE comment explaining why `fmt.token_dim()` cannot be used in `reshape_partial_chunk`), which should help future contributors understand the constraint. A proper fix would likely involve introducing dedicated layerwise MemoryFormat variants  worth tracking as a follow-up issue.

**Comment by ianliuy** [comment]
Hi @ApostaC @DongDongJu  this PR is ready for review. Quick summary:

- Follow-up to #2751 (layerwise remote backend fix), adding missing test coverage and bug fixes found during review
- All review feedback from gemini-code-assist has been addressed (multi-group reshape support, assertValueError, MemoryFormat scope clarification)
- Rebased onto latest `dev`, CI running

The PR needs 2 code owner approvals to merge. Would appreciate a look when you get a chance. Thanks!
